File: validate.py
#!/usr/bin/env python3
"""
project.md validator — checks conformance with the project.md specification.

Usage:
    python3 validate.py [path/to/project.md]

If no path given, looks for project.md in the current directory.
"""

import re
import sys
import os

try:
    import yaml
    HAS_YAML = True
except ImportError:
    HAS_YAML = False


REQUIRED_FRONTMATTER = ["project", "status", "created", "updated", "owner"]
VALID_STATUSES = {"define", "design", "build", "verify", "ship", "paused", "archived"}
REQUIRED_SECTIONS = [
    "What This Is",
    "Core Value",
    "Requirements",
    "Current State",
    "Key Decisions",
    "Tasks",
]
RECOMMENDED_SECTIONS = [
    "Context",
    "Constraints",
    "Architecture",
    "Discoveries",
    "References",
    "Session Log",
]
REQUIREMENT_TIERS = ["Validated", "Active", "Out of Scope"]


class ValidationResult:
    def __init__(self):
        self.errors = []
        self.warnings = []
        self.info = []

    def error(self, msg):
        self.errors.append(msg)

    def warn(self, msg):
        self.warnings.append(msg)

    def ok(self, msg):
        self.info.append(msg)

    @property
    def passed(self):
        return len(self.errors) == 0

    def report(self):
        lines = []
        lines.append("=" * 60)
        lines.append("  project.md Validation Report")
        lines.append("=" * 60)
        lines.append("")

        if self.info:
            lines.append("PASS:")
            for msg in self.info:
                lines.append(f"  + {msg}")
            lines.append("")

        if self.warnings:
            lines.append("WARNINGS:")
            for msg in self.warnings:
                lines.append(f"  ! {msg}")
            lines.append("")

        if self.errors:
            lines.append("ERRORS:")
            for msg in self.errors:
                lines.append(f"  x {msg}")
            lines.append("")

        lines.append("-" * 60)
        if self.passed:
            lines.append(f"  RESULT: PASS ({len(self.warnings)} warnings)")
        else:
            lines.append(f"  RESULT: FAIL ({len(self.errors)} errors, {len(self.warnings)} warnings)")
        lines.append("-" * 60)

        return "\n".join(lines)


def parse_frontmatter(content):
    """Extract YAML frontmatter from markdown content."""
    match = re.match(r"^---\s*\n(.*?)\n---\s*\n", content, re.DOTALL)
    if not match:
        return None, content

    raw = match.group(1)
    body = content[match.end():]

    if HAS_YAML:
        try:
            return yaml.safe_load(raw), body
        except yaml.YAMLError:
            return None, body
    else:
        # Basic parsing without PyYAML
        fm = {}
        for line in raw.split("\n"):
            line = line.strip()
            if ":" in line and not line.startswith("#"):
                key, _, value = line.partition(":")
                key = key.strip()
                value = value.strip().strip('"').strip("'")
                if value.startswith("[") and value.endswith("]"):
                    # Basic array parsing
                    value = [v.strip().strip('"').strip("'")
                             for v in value[1:-1].split(",") if v.strip()]
                fm[key] = value
        return fm, body


def find_sections(body):
    """Find all ## headings in the document."""
    sections = {}
    for match in re.finditer(r"^##\s+(.+)$", body, re.MULTILINE):
        name = match.group(1).strip()
        sections[name] = match.start()
    return sections


def validate(filepath):
    result = ValidationResult()

    if not os.path.exists(filepath):
        result.error(f"File not found: {filepath}")
        return result

    with open(filepath, "r", encoding="utf-8") as f:
        content = f.read()

    # --- Frontmatter ---
    fm, body = parse_frontmatter(content)

    if fm is None:
        result.error("Missing or invalid YAML frontmatter (must start with ---)")
        return result

    result.ok("YAML frontmatter present")

    for field in REQUIRED_FRONTMATTER:
        if field not in fm:
            result.error(f"Missing required frontmatter field: {field}")
        elif not fm[field]:
            result.error(f"Empty required frontmatter field: {field}")
        else:
            result.ok(f"Frontmatter field '{field}' = {fm[field]}")

    # Validate status
    if "status" in fm and fm["status"]:
        if fm["status"] not in VALID_STATUSES:
            result.error(f"Invalid status '{fm['status']}'. Must be one of: {', '.join(sorted(VALID_STATUSES))}")
        else:
            result.ok(f"Status '{fm['status']}' is valid")

    # Validate dates
    for date_field in ["created", "updated"]:
        if date_field in fm and fm[date_field]:
            date_str = str(fm[date_field])
            if not re.match(r"^\d{4}-\d{2}-\d{2}$", date_str):
                result.warn(f"Field '{date_field}' is not ISO 8601 format (YYYY-MM-DD): {date_str}")
            else:
                result.ok(f"Field '{date_field}' is valid ISO 8601: {date_str}")

    # --- Sections ---
    sections = find_sections(body)

    for section in REQUIRED_SECTIONS:
        if section in sections:
            result.ok(f"Required section found: {section}")
        else:
            result.error(f"Missing required section: ## {section}")

    for section in RECOMMENDED_SECTIONS:
        if section in sections:
            result.ok(f"Recommended section found: {section}")
        else:
            result.warn(f"Missing recommended section: ## {section}")

    # --- Requirements lifecycle ---
    if "Requirements" in sections:
        has_validated = "Validated" in body
        has_active = "Active" in body
        has_out_of_scope = "Out of Scope" in body

        if has_validated:
            result.ok("Requirements tier found: Validated")
        else:
            result.warn("Requirements tier missing: Validated (expected for conformance)")

        if has_active:
            result.ok("Requirements tier found: Active")
        else:
            result.error("Requirements tier missing: Active")

        if has_out_of_scope:
            result.ok("Requirements tier found: Out of Scope")
        else:
            result.error("Requirements tier missing: Out of Scope")

    # --- Tasks ---
    tasks_pending = len(re.findall(r"^- \[ \]", body, re.MULTILINE))
    tasks_done = len(re.findall(r"^- \[x\]", body, re.MULTILINE))
    tasks_blocked = len(re.findall(r"^- \[!\]", body, re.MULTILINE))
    total_tasks = tasks_pending + tasks_done + tasks_blocked

    if total_tasks == 0:
        result.warn("No tasks found (expected checkbox syntax: - [ ], - [x], - [!])")
    else:
        result.ok(f"Tasks found: {total_tasks} total ({tasks_done} done, {tasks_pending} pending, {tasks_blocked} blocked)")

    # --- Key Decisions ---
    if "Key Decisions" in sections:
        # Check for table format
        table_match = re.search(r"\| Decision.*\|.*Rationale.*\|.*Outcome.*\|", body)
        if table_match:
            result.ok("Key Decisions table has correct columns (Decision, Rationale, Outcome)")
        else:
            result.warn("Key Decisions section exists but table may be missing columns (expected: Decision | Rationale | Outcome)")

        # Check for outcome tracking
        has_good = "Good" in body or "✓" in body
        has_revisit = "Revisit" in body or "⚠️" in body
        has_pending = "Pending" in body or "—" in body
        if has_good or has_revisit or has_pending:
            result.ok("Decision outcome tracking detected")

    # --- Core Value ---
    if "Core Value" in sections:
        # Check it's roughly one sentence (not too long)
        cv_start = sections["Core Value"]
        next_section = None
        for name, pos in sections.items():
            if pos > cv_start and name != "Core Value":
                if next_section is None or pos < next_section:
                    next_section = pos

        if next_section:
            cv_text = body[cv_start:next_section].strip()
            # Remove the heading
            cv_text = re.sub(r"^## Core Value\s*", "", cv_text).strip()
            cv_lines = [l for l in cv_text.split("\n") if l.strip() and not l.strip().startswith("<!--")]
            if cv_lines:
                cv_content = " ".join(cv_lines)
                sentence_count = len(re.findall(r"[.!?]+", cv_content))
                if sentence_count > 3:
                    result.warn(f"Core Value has {sentence_count} sentences — should be ~1 sentence")
                else:
                    result.ok("Core Value is concise")

    # --- What This Is ---
    if "What This Is" in sections:
        wti_start = sections["What This Is"]
        next_section = None
        for name, pos in sections.items():
            if pos > wti_start and name != "What This Is":
                if next_section is None or pos < next_section:
                    next_section = pos

        if next_section:
            wti_text = body[wti_start:next_section].strip()
            wti_text = re.sub(r"^## What This Is\s*", "", wti_text).strip()
            wti_lines = [l for l in wti_text.split("\n") if l.strip() and not l.strip().startswith("<!--")]
            if wti_lines:
                wti_content = " ".join(wti_lines)
                if len(wti_content) > 500:
                    result.warn(f"'What This Is' is {len(wti_content)} chars — should be 2-3 sentences")
                else:
                    result.ok("'What This Is' length is appropriate")

    # --- Current State ---
    if "Current State" in sections:
        cs_markers = ["Phase:", "Last completed:", "Next action:", "Blockers:"]
        found = sum(1 for m in cs_markers if m in body)
        if found >= 3:
            result.ok(f"Current State has {found}/{len(cs_markers)} expected fields")
        else:
            result.warn(f"Current State only has {found}/{len(cs_markers)} expected fields (Phase, Last completed, Next action, Blockers)")

    return result


def main():
    if len(sys.argv) > 1:
        filepath = sys.argv[1]
    else:
        filepath = "project.md"

    result = validate(filepath)
    print(result.report())
    sys.exit(0 if result.passed else 1)


if __name__ == "__main__":
    main()

File: src/validate.rs
//! Conformance validation for project.md files.
//!
//! Checks that a project.md file conforms to the expected structure and schema.

use crate::decisions::parse_decisions;
use crate::project::Project;
use crate::requirements::parse_requirements;
use crate::sections::find_section;
use crate::tasks::parse_all_tasks;
use chrono::NaiveDate;

/// Validation result with categorized messages.
#[derive(Debug, Clone)]
pub struct ValidationResult {
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
    pub info: Vec<String>,
}

impl ValidationResult {
    /// Returns true if there are no errors.
    pub fn passed(&self) -> bool {
        self.errors.is_empty()
    }

    /// Format a human-readable validation report.
    pub fn report(&self) -> String {
        let mut out = String::new();

        if self.passed() {
            out.push_str("PASS: Project conforms to spec\n");
        } else {
            out.push_str("FAIL: Project has validation errors\n");
        }

        if !self.errors.is_empty() {
            out.push_str(&format!("\nErrors ({}):\n", self.errors.len()));
            for e in &self.errors {
                out.push_str(&format!("  ✗ {}\n", e));
            }
        }

        if !self.warnings.is_empty() {
            out.push_str(&format!("\nWarnings ({}):\n", self.warnings.len()));
            for w in &self.warnings {
                out.push_str(&format!("  ⚠ {}\n", w));
            }
        }

        if !self.info.is_empty() {
            out.push_str(&format!("\nInfo ({}):\n", self.info.len()));
            for i in &self.info {
                out.push_str(&format!("  ℹ {}\n", i));
            }
        }

        out
    }
}

/// Validate a project.md file for conformance with the spec.
pub fn validate(project: &Project) -> ValidationResult {
    let mut result = ValidationResult {
        errors: Vec::new(),
        warnings: Vec::new(),
        info: Vec::new(),
    };

    validate_frontmatter(project, &mut result);
    validate_required_sections(project, &mut result);
    validate_recommended_sections(project, &mut result);
    validate_requirements(project, &mut result);
    validate_tasks(project, &mut result);
    validate_decisions(project, &mut result);
    validate_core_value(project, &mut result);
    validate_current_state(project, &mut result);

    result
}

/// Check frontmatter required fields, status enum, and ISO 8601 dates.
fn validate_frontmatter(project: &Project, result: &mut ValidationResult) {
    let fm = &project.frontmatter;

    // Required fields
    if fm.project.trim().is_empty() {
        result
            .errors
            .push("Frontmatter: 'project' field is empty".into());
    }
    if fm.owner.trim().is_empty() {
        result
            .errors
            .push("Frontmatter: 'owner' field is empty".into());
    }

    // Check dates are reasonable (not default/epoch)
    let min_date = NaiveDate::from_ymd_opt(2020, 1, 1).unwrap();
    if fm.created < min_date {
        result.errors.push(format!(
            "Frontmatter: 'created' date {} seems invalid (before 2020)",
            fm.created
        ));
    }
    if fm.updated < min_date {
        result.errors.push(format!(
            "Frontmatter: 'updated' date {} seems invalid (before 2020)",
            fm.updated
        ));
    }
    if fm.updated < fm.created {
        result.errors.push(format!(
            "Frontmatter: 'updated' ({}) is before 'created' ({})",
            fm.updated, fm.created
        ));
    }

    // Status is already validated by serde deserialization, but check it's a known value
    result.info.push(format!("Status: {}", fm.status));

    // Optional field info
    if fm.agent.is_none() {
        result.info.push("No agent specified in frontmatter".into());
    }
    if fm.tags.is_empty() {
        result.info.push("No tags specified".into());
    }
}

/// Check that all required sections are present.
fn validate_required_sections(project: &Project, result: &mut ValidationResult) {
    let required = [
        "What This Is",
        "Core Value",
        "Requirements",
        "Current State",
        "Key Decisions",
        "Tasks",
    ];

    for name in &required {
        if find_section(&project.sections, name).is_none() {
            result
                .errors
                .push(format!("Missing required section: '{}'", name));
        }
    }
}

/// Check that recommended sections are present (warnings only).
fn validate_recommended_sections(project: &Project, result: &mut ValidationResult) {
    let recommended = [
        "Context",
        "Constraints",
        "Architecture",
        "Discoveries",
        "References",
        "Session Log",
    ];

    for name in &recommended {
        if find_section(&project.sections, name).is_none() {
            result
                .warnings
                .push(format!("Missing recommended section: '{}'", name));
        }
    }
}

/// Check requirement tiers: all three present.
fn validate_requirements(project: &Project, result: &mut ValidationResult) {
    if let Some(req_section) = find_section(&project.sections, "Requirements") {
        let reqs = parse_requirements(&req_section.content);

        if reqs.validated.is_empty() {
            result
                .warnings
                .push("Requirements: No validated requirements found".into());
        }
        if reqs.active.is_empty() {
            result
                .warnings
                .push("Requirements: No active requirements found".into());
        }

        // Check that tier subsection headers exist in the content
        let content = &req_section.content;
        if !content.contains("### Validated") && !content.contains("### validated") {
            result
                .errors
                .push("Requirements: Missing '### Validated' tier".into());
        }
        if !content.contains("### Active") && !content.contains("### active") {
            result
                .errors
                .push("Requirements: Missing '### Active' tier".into());
        }
        if !content.contains("### Out of Scope") && !content.contains("### out of scope") {
            result
                .errors
                .push("Requirements: Missing '### Out of Scope' tier".into());
        }

        result.info.push(format!(
            "Requirements: {} validated, {} active, {} out of scope",
            reqs.validated.len(),
            reqs.active.len(),
            reqs.out_of_scope.len()
        ));
    }
}

/// Check task syntax: count - [ ], - [x], - [!].
fn validate_tasks(project: &Project, result: &mut ValidationResult) {
    if let Some(tasks_section) = find_section(&project.sections, "Tasks") {
        let all_tasks = parse_all_tasks(&project.sections);

        let mut total_pending = 0;
        let mut total_done = 0;
        let mut total_blocked = 0;

        for (_phase, tasks) in &all_tasks {
            for task in tasks {
                match task.status {
                    crate::tasks::TaskStatus::Pending => total_pending += 1,
                    crate::tasks::TaskStatus::Done => total_done += 1,
                    crate::tasks::TaskStatus::Blocked => total_blocked += 1,
                }
            }
        }

        if all_tasks.is_empty() {
            result.warnings.push("Tasks: No task phases found".into());
        }

        result.info.push(format!(
            "Tasks: {} done, {} pending, {} blocked (across {} phases)",
            total_done,
            total_pending,
            total_blocked,
            all_tasks.len()
        ));

        // Check for phase headers
        if !tasks_section.content.contains("### Phase:") {
            result
                .warnings
                .push("Tasks: No '### Phase:' subsections found".into());
        }
    }
}

/// Check decisions table has Decision, Rationale, Outcome columns.
fn validate_decisions(project: &Project, result: &mut ValidationResult) {
    if let Some(dec_section) = find_section(&project.sections, "Key Decisions") {
        let decisions = parse_decisions(&dec_section.content);

        if decisions.is_empty() {
            result
                .warnings
                .push("Key Decisions: No decisions found in table".into());
        }

        // Check for table header
        let content = &dec_section.content;
        if !content.contains("Decision")
            || !content.contains("Rationale")
            || !content.contains("Outcome")
        {
            result.errors.push(
                "Key Decisions: Table missing required columns (Decision, Rationale, Outcome)"
                    .into(),
            );
        }

        let good = decisions
            .iter()
            .filter(|d| d.outcome == crate::decisions::Outcome::Good)
            .count();
        let pending = decisions
            .iter()
            .filter(|d| d.outcome == crate::decisions::Outcome::Pending)
            .count();
        let revisit = decisions
            .iter()
            .filter(|d| d.outcome == crate::decisions::Outcome::Revisit)
            .count();

        result.info.push(format!(
            "Decisions: {} total ({} good, {} pending, {} revisit)",
            decisions.len(),
            good,
            pending,
            revisit
        ));
    }
}

/// Check Core Value conciseness (roughly 1-3 sentences).
fn validate_core_value(project: &Project, result: &mut ValidationResult) {
    if let Some(section) = find_section(&project.sections, "Core Value") {
        let content = section.content.trim();
        if content.is_empty() {
            result.errors.push("Core Value: Section is empty".into());
            return;
        }

        // Count sentences (rough heuristic: count sentence-ending punctuation)
        let sentence_count = content
            .chars()
            .filter(|c| *c == '.' || *c == '!' || *c == '?')
            .count();

        if sentence_count == 0 {
            result.warnings.push(
                "Core Value: No sentence-ending punctuation found — should be 1-3 sentences".into(),
            );
        } else if sentence_count > 3 {
            result.warnings.push(format!(
                "Core Value: Found ~{} sentences — should be concise (1-3 sentences)",
                sentence_count
            ));
        }

        if content.len() > 500 {
            result.warnings.push(format!(
                "Core Value: {} characters — consider being more concise",
                content.len()
            ));
        }
    }
}

/// Check Current State completeness.
fn validate_current_state(project: &Project, result: &mut ValidationResult) {
    if let Some(section) = find_section(&project.sections, "Current State") {
        let content = &section.content;

        let required_fields = ["Phase", "Last completed", "Next action", "Blockers"];
        for field in &required_fields {
            let pattern = format!("**{}:**", field);
            if !content.contains(&pattern) {
                result
                    .errors
                    .push(format!("Current State: Missing required field '{}'", field));
            }
        }

        // Check if Phase value is set
        if let Some(pos) = content.find("**Phase:**") {
            let after = &content[pos + "**Phase:**".len()..];
            let value = after.lines().next().unwrap_or("").trim();
            if value.is_empty() {
                result
                    .errors
                    .push("Current State: Phase value is empty".into());
            }
        }

        // Check if Blockers is explicitly stated
        if let Some(pos) = content.find("**Blockers:**") {
            let after = &content[pos + "**Blockers:**".len()..];
            let value = after.lines().next().unwrap_or("").trim();
            if value.is_empty() {
                result.warnings.push(
                    "Current State: Blockers field is empty — specify 'None' if no blockers".into(),
                );
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    const VALID_PROJECT: &str = r#"---
project: "Test Project"
status: build
created: 2026-01-01
updated: 2026-04-29
owner: "Test Owner"
agent: "Test Agent"
tags: [test]
---

## What This Is

A test project for validation.

## Core Value

Fast, reliable testing.

## Requirements

### Validated

- ✓ Feature A — v0.1

### Active

- [ ] Feature B

### Out of Scope

- Feature C — not needed

## Context

Some context.

## Constraints

Some constraints.

## Current State

**Phase:** build
**Last completed:** Task 1
**In progress:** Task 2
**Next action:** Do the thing
**Blockers:** None
**Notes:** Some notes

## Architecture

Some architecture.

## Key Decisions

| Decision | Rationale | Outcome |
|----------|-----------|---------|
| Use Rust | Fast and safe | ✓ Good |

## Tasks

### Phase: BUILD

- [x] Task 1: Setup
- [ ] Task 2: Implementation

## Discoveries

Some discoveries.

## References

Some references.

## Session Log

- **2026-01-01** — Project started.
"#;

    #[test]
    fn test_valid_project_passes() {
        let project = Project::from_str(VALID_PROJECT, Path::new("/tmp/test.md")).unwrap();
        let result = validate(&project);
        assert!(
            result.passed(),
            "Expected pass, got errors: {:?}",
            result.errors
        );
    }

    #[test]
    fn test_missing_required_sections() {
        let content = r#"---
project: "Test"
status: build
created: 2026-01-01
updated: 2026-04-29
owner: "Owner"
---

## What This Is

Something.
"#;
        let project = Project::from_str(content, Path::new("/tmp/test.md")).unwrap();
        let result = validate(&project);
        assert!(!result.passed());
        assert!(result.errors.iter().any(|e| e.contains("Core Value")));
        assert!(result.errors.iter().any(|e| e.contains("Requirements")));
        assert!(result.errors.iter().any(|e| e.contains("Current State")));
        assert!(result.errors.iter().any(|e| e.contains("Key Decisions")));
        assert!(result.errors.iter().any(|e| e.contains("Tasks")));
    }

    #[test]
    fn test_missing_recommended_sections() {
        let content = r#"---
project: "Test"
status: build
created: 2026-01-01
updated: 2026-04-29
owner: "Owner"
---

## What This Is

Something.

## Core Value

Short value.

## Requirements

### Validated

- ✓ A — v0.1

### Active

- [ ] B

### Out of Scope

- C — nope

## Current State

**Phase:** build
**Last completed:** Task 1
**Next action:** Do thing
**Blockers:** None

## Key Decisions

| Decision | Rationale | Outcome |
|----------|-----------|---------|
| A | B | ✓ Good |

## Tasks

### Phase: BUILD

- [ ] Task 1: Something
"#;
        let project = Project::from_str(content, Path::new("/tmp/test.md")).unwrap();
        let result = validate(&project);
        // Should have warnings for missing recommended sections
        assert!(result.warnings.iter().any(|w| w.contains("Context")));
        assert!(result.warnings.iter().any(|w| w.contains("Constraints")));
        assert!(result.warnings.iter().any(|w| w.contains("Architecture")));
        assert!(result.warnings.iter().any(|w| w.contains("Discoveries")));
        assert!(result.warnings.iter().any(|w| w.contains("References")));
        assert!(result.warnings.iter().any(|w| w.contains("Session Log")));
    }

    #[test]
    fn test_empty_project_name() {
        let content = r#"---
project: ""
status: build
created: 2026-01-01
updated: 2026-04-29
owner: "Owner"
---

## What This Is

Something.

## Core Value

Fast.

## Requirements

### Validated
- ✓ A — v0.1
### Active
- [ ] B
### Out of Scope
- C — nope

## Current State

**Phase:** build
**Last completed:** T1
**Next action:** T2
**Blockers:** None

## Key Decisions

| Decision | Rationale | Outcome |
|----------|-----------|---------|
| A | B | ✓ Good |

## Tasks

### Phase: BUILD
- [ ] T1
"#;
        let project = Project::from_str(content, Path::new("/tmp/test.md")).unwrap();
        let result = validate(&project);
        assert!(result
            .errors
            .iter()
            .any(|e| e.contains("'project' field is empty")));
    }

    #[test]
    fn test_updated_before_created() {
        let content = r#"---
project: "Test"
status: build
created: 2026-04-29
updated: 2026-01-01
owner: "Owner"
---

## What This Is

Something.

## Core Value

Fast.

## Requirements

### Validated
- ✓ A — v0.1
### Active
- [ ] B
### Out of Scope
- C — nope

## Current State

**Phase:** build
**Last completed:** T1
**Next action:** T2
**Blockers:** None

## Key Decisions

| Decision | Rationale | Outcome |
|----------|-----------|---------|
| A | B | ✓ Good |

## Tasks

### Phase: BUILD
- [ ] T1
"#;
        let project = Project::from_str(content, Path::new("/tmp/test.md")).unwrap();
        let result = validate(&project);
        assert!(result
            .errors
            .iter()
            .any(|e| e.contains("'updated'") && e.contains("before 'created'")));
    }

    #[test]
    fn test_validation_report_format() {
        let project = Project::from_str(VALID_PROJECT, Path::new("/tmp/test.md")).unwrap();
        let result = validate(&project);
        let report = result.report();
        assert!(report.contains("PASS"));
    }

    #[test]
    fn test_fail_report_format() {
        let content = r#"---
project: "Test"
status: build
created: 2026-01-01
updated: 2026-04-29
owner: "Owner"
---
"#;
        let project = Project::from_str(content, Path::new("/tmp/test.md")).unwrap();
        let result = validate(&project);
        let report = result.report();
        assert!(report.contains("FAIL"));
        assert!(report.contains("Errors"));
    }

    #[test]
    fn test_current_state_missing_fields() {
        let content = r#"---
project: "Test"
status: build
created: 2026-01-01
updated: 2026-04-29
owner: "Owner"
---

## What This Is

Something.

## Core Value

Fast.

## Requirements

### Validated
- ✓ A — v0.1
### Active
- [ ] B
### Out of Scope
- C — nope

## Current State

**Phase:** build

## Key Decisions

| Decision | Rationale | Outcome |
|----------|-----------|---------|
| A | B | ✓ Good |

## Tasks

### Phase: BUILD
- [ ] T1
"#;
        let project = Project::from_str(content, Path::new("/tmp/test.md")).unwrap();
        let result = validate(&project);
        assert!(result.errors.iter().any(|e| e.contains("Last completed")));
        assert!(result.errors.iter().any(|e| e.contains("Next action")));
        assert!(result.errors.iter().any(|e| e.contains("Blockers")));
    }

    #[test]
    fn test_validate_example_cli_tool() {
        let content = std::fs::read_to_string(
            std::path::Path::new(env!("CARGO_MANIFEST_DIR")).join("example-cli-tool.md"),
        )
        .unwrap();
        let project = Project::from_str(&content, Path::new("/tmp/test.md")).unwrap();
        let result = validate(&project);
        // The example should pass validation
        assert!(
            result.passed(),
            "Example should pass, got errors: {:?}",
            result.errors
        );
    }
}
